### PR TITLE
[SDK-574] Add ResourceEncryptionService

### PR DESCRIPTION
**Pull request by bitPogo**
## Description

This introduces a new Service - the `ResourceEncryptionService`. Its purpose is move the encryption logic out of the RecordService. Also the DecryptedRecordMapper is not longer needed and was removed.

## Motivation and Context

The RecordService is still too huge. This a step to make it smaller.
The PR is part of [SDK-574](https://gesundheitscloud.atlassian.net/browse/SDK-574) and depends on #122.

## How is it being implemented?

As mentioned it adds a new service alongside with its interface. It moves the en/decryption logic out of the RecordService and removes the DecryptedRecordMapper.

## How Has This Been Tested?

A new test set was added to cover the changes. Also all tests of effected classes had been changed accordingly. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated the changelog accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: sdk-core/src/main/java/care/data4life/sdk/record/RecordContract.kt [hunk 1]
```diff
@@ -30,8 +30,7 @@ import io.reactivex.Completable
 import io.reactivex.Single
 import org.threeten.bp.LocalDate
 
-interface RecordContract {
-
+class RecordContract {
```
Comment (wmontwe) [inline sdk-core/src/main/java/care/data4life/sdk/record/RecordContract.kt:33]: Why was this changed to a class?

File: sdk-core/src/main/java/care/data4life/sdk/network/model/NetworkModelContract.kt [hunk 2]
```diff
@@ -20,51 +20,12 @@ import care.data4life.crypto.GCKey
 import care.data4life.sdk.data.DataResource
 import care.data4life.sdk.fhir.Fhir3Resource
 import care.data4life.sdk.fhir.Fhir4Resource
-import care.data4life.sdk.lang.CoreRuntimeException
 import care.data4life.sdk.lang.DataValidationException
 import care.data4life.sdk.tag.Annotations
 import care.data4life.sdk.tag.EncryptedTagsAndAnnotations
 import care.data4life.sdk.tag.Tags
 
 class NetworkModelContract {
```
Comment (wmontwe) [inline sdk-core/src/main/java/care/data4life/sdk/network/model/NetworkModelContract.kt:28]: I notice that contracts are classes, why that?

Comment (bitPogo) [reply sdk-core/src/main/java/care/data4life/sdk/network/model/NetworkModelContract.kt:28]: EncryptionService and Version are `internal`.

File: sdk-core/src/main/java/care/data4life/sdk/record/RecordContract.kt (hunk 1 above)
Comment (bitPogo) [reply sdk-core/src/main/java/care/data4life/sdk/record/RecordContract.kt:33]: Good catch..this one seemed to slipped while rebasing.

Comment (bitPogo) [reply sdk-core/src/main/java/care/data4life/sdk/record/RecordContract.kt]: done

File: sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+import care.data4life.sdk.lang.CoreRuntimeException
+import care.data4life.sdk.lang.DataValidationException
+import care.data4life.sdk.model.ModelContract
+import care.data4life.sdk.model.ModelContract.ModelVersion.Companion.CURRENT
+import care.data4life.sdk.tag.Annotations
+import care.data4life.sdk.tag.TaggingContract
+import care.data4life.sdk.tag.Tags
+import care.data4life.sdk.wrapper.WrapperContract
+
+class RecordEncryptionService(
```
Comment (alessioborraccino) [inline sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt:36]: is final class a thing in Kotlin?

File: sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+        if (encryptedResource.isEmpty()) {
+            throw CoreRuntimeException.InternalFailure()
+        }
+
+        if (!modelVersion.isModelVersionSupported(version)) {
+            throw DataValidationException.ModelVersionNotSupported("Please update SDK to latest version!")
+        }
+    }
+
+    private fun getCommonKey(commonKeyId: String, userId: String): GCKey {
```
Comment (alessioborraccino) [inline sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt:258]: Naming is a bit confusing to me, we have 
`fetchCommonKey()` which returns a pair of GCKey and id
`getCommonKey()` which returns a GCKey 

and i wouldnt know which one refers to which by reading method names

File: sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+        return if (commonKeyStored) {
+            cryptoService.getCommonKeyById(commonKeyId)
+        } else {
+            // TODO: This should be in a different Service
+            val commonKeyResponse = apiService.fetchCommonKey(
+                alias,
+                userId,
+                commonKeyId
+            ).blockingGet()
+
```
Comment (alessioborraccino) [inline sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt:269]: I see the TODO :) we created a `CommonKeyService` to wrap up all the CommonKey shenanigans

File: sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+        val commonKey = getCommonKey(encryptedRecord.commonKeyId, userId)
+        val (dataKey, attachmentKey) = decryptKeys(
+            commonKey,
+            encryptedRecord.encryptedDataKey,
+            encryptedRecord.encryptedAttachmentsKey
+        )
+
+        return buildRecord(
+            encryptedRecord.identifier,
+            fhirService.decryptResource(dataKey, tags, encryptedRecord.encryptedBody),
```
Comment (alessioborraccino) [inline sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt:321]: is there any way that you can avoid using the `fhirService` here? seems very out of place to me, If this is a record(resource)cryptor

**Review by alessioborraccino — CHANGES_REQUESTED** [review]

File: sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt (hunk 4 above)
Comment (bitPogo) [reply sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt:258]: Have suggestions for better names?

File: sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt (hunk 3 above)
Comment (bitPogo) [reply sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt:36]: By default, all classes in Kotlin are final.

File: sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt (hunk 5 above)
Comment (bitPogo) [reply sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt:269]: Indeed...the whole key stuff should be encapsulated somewhere else...but to do it right now cause more trouble than it helps. However this will happen once Koin (DI) is in place.

File: sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt (hunk 6 above)
Comment (bitPogo) [reply sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt:321]: No, we cannot avoid that for now. The FhirService is actually more a ResourceCryptoService. We'll come to this once we bubbled up to the corresponding part in the SDK.

File: sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt (hunk 4 above)
Comment (bitPogo) [reply sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt:258]: On a second look...and since its technically a private method...I guess it maps to the `getCommonKey` method you mentioned. Sure, the signature is different in terms of the parameter but the name and return value matches.

File: sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt (hunk 6 above)
Comment (bitPogo) [reply sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt:321]: I think the root cause, is more like, then we should slice out this whole mapping record stuff out (same goes for the `fromRecord` method - it will live in the future somewhere else). However I did not it here since it causes much less effort, when finally Koin is in place.

Comment (bitPogo) [reply sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt:321]: Totally forgot, we'll get rid of the entire mappings of the resources internally, since they will be wrapped. This will in total improve the whole resource handling as well as the readability, but this something which will be seen later in time, since there’s  still much more in queue before.
However I made a follow up to address the overall naming issues - #128.

File: sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+    private val apiService: ApiService,
+    private val taggingService: TaggingContract.Service,
+    private val tagEncryptionService: TaggingContract.EncryptionService,
+    private val guard: NetworkModelContract.LimitGuard,
+    private val cryptoService: CryptoContract.Service,
+    private val fhirService: FhirContract.Service,
+    private val dateTimeFormatter: WrapperContract.DateTimeFormatter,
+    private val modelVersion: ModelContract.ModelVersion
+) : NetworkModelContract.EncryptionService {
+    private fun <T : Fhir3Resource> buildFhir3Record(
```
Comment (wmontwe) [inline sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt:47]: RecordEncryptionService is now also responsible for mapping, which blows up the responsibility. Ideally it could delegate that work.

**Review by wmontwe — APPROVED** [review]
RecordEncryptionService is too heavy but not a deal breaker.

**Review by alessioborraccino — APPROVED** [review]
